File: test_common_equivalence.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>
#include <cmath>
#include <iomanip>

// Include the original LightGBM common.h header
#include "../include/LightGBM/utils/common.h"

using namespace LightGBM::Common;

void test_string_functions() {
    std::cout << "\n=== Testing String Functions ===\n";
    
    // Test tolower
    std::cout << "tolower('A'): " << tolower('A') << std::endl;
    std::cout << "tolower('Z'): " << tolower('Z') << std::endl;
    std::cout << "tolower('a'): " << tolower('a') << std::endl;
    std::cout << "tolower('1'): " << tolower('1') << std::endl;
    
    // Test Trim
    std::cout << "Trim('  hello  '): '" << Trim("  hello  ") << "'" << std::endl;
    std::cout << "Trim('\\t\\n hello \\r\\n'): '" << Trim("\t\n hello \r\n") << "'" << std::endl;
    std::cout << "Trim(''): '" << Trim("") << "'" << std::endl;
    
    // Test RemoveQuotationSymbol
    std::cout << "RemoveQuotationSymbol('\"hello\"'): '" << RemoveQuotationSymbol("\"hello\"") << "'" << std::endl;
    std::cout << "RemoveQuotationSymbol(\"'world'\"): '" << RemoveQuotationSymbol("'world'") << "'" << std::endl;
    
    // Test StartsWith
    std::cout << "StartsWith('hello world', 'hello'): " << std::boolalpha << StartsWith("hello world", "hello") << std::endl;
    std::cout << "StartsWith('hello world', 'world'): " << std::boolalpha << StartsWith("hello world", "world") << std::endl;
    
    // Test Split
    std::vector<std::string> split_result = Split("a,b,c", ',');
    std::cout << "Split('a,b,c', ','): ";
    for (const auto& s : split_result) {
        std::cout << "'" << s << "' ";
    }
    std::cout << std::endl;
    
    split_result = Split("a,,b", ',');
    std::cout << "Split('a,,b', ','): ";
    for (const auto& s : split_result) {
        std::cout << "'" << s << "' ";
    }
    std::cout << std::endl;
    
    // Test SplitLines
    std::vector<std::string> lines_result = SplitLines("line1\nline2\r\nline3");
    std::cout << "SplitLines('line1\\nline2\\r\\nline3'): ";
    for (const auto& s : lines_result) {
        std::cout << "'" << s << "' ";
    }
    std::cout << std::endl;
}

void test_numeric_functions() {
    std::cout << "\n=== Testing Numeric Functions ===\n";
    
    // Test Atoi
    int int_result;
    const char* remaining = Atoi("123abc", &int_result);
    std::cout << "Atoi('123abc'): value=" << int_result << ", remaining='" << remaining << "'" << std::endl;
    
    remaining = Atoi("-456", &int_result);
    std::cout << "Atoi('-456'): value=" << int_result << ", remaining='" << remaining << "'" << std::endl;
    
    // Test Atof
    double double_result;
    remaining = Atof("123.45", &double_result);
    std::cout << "Atof('123.45'): value=" << std::fixed << std::setprecision(6) << double_result << std::endl;
    
    remaining = Atof("1.23e-4", &double_result);
    std::cout << "Atof('1.23e-4'): value=" << std::scientific << std::setprecision(6) << double_result << std::endl;
    
    remaining = Atof("nan", &double_result);
    std::cout << "Atof('nan'): value=" << (std::isnan(double_result) ? "NaN" : std::to_string(double_result)) << std::endl;
    
    remaining = Atof("inf", &double_result);
    std::cout << "Atof('inf'): value=" << (std::isinf(double_result) ? "Inf" : std::to_string(double_result)) << std::endl;
    
    // Test AtoiAndCheck
    std::cout << "AtoiAndCheck('123'): " << std::boolalpha << AtoiAndCheck("123", &int_result) << std::endl;
    std::cout << "AtoiAndCheck('123abc'): " << std::boolalpha << AtoiAndCheck("123abc", &int_result) << std::endl;
    
    // Test AtofAndCheck
    std::cout << "AtofAndCheck('123.45'): " << std::boolalpha << AtofAndCheck("123.45", &double_result) << std::endl;
    std::cout << "AtofAndCheck('123.45abc'): " << std::boolalpha << AtofAndCheck("123.45abc", &double_result) << std::endl;
}

void test_array_functions() {
    std::cout << "\n=== Testing Array Functions ===\n";
    
    // Test ArrayCast
    std::vector<int> int_vec = {1, 2, 3, 4, 5};
    std::vector<double> double_vec = ArrayCast<int, double>(int_vec);
    std::cout << "ArrayCast<int, double>([1,2,3,4,5]): ";
    for (double d : double_vec) {
        std::cout << d << " ";
    }
    std::cout << std::endl;
    
    // Test StringToArray
    std::vector<int> str_to_array = StringToArray<int>("1,2,3,4,5", ',');
    std::cout << "StringToArray<int>('1,2,3,4,5', ','): ";
    for (int i : str_to_array) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
    
    // Test Join
    std::vector<int> join_vec = {10, 20, 30, 40};
    std::string joined = Join(join_vec, ", ");
    std::cout << "Join([10,20,30,40], ', '): '" << joined << "'" << std::endl;
    
    std::string joined_range = Join(join_vec, 1, 3, " | ");
    std::cout << "Join([10,20,30,40], 1, 3, ' | '): '" << joined_range << "'" << std::endl;
}

void test_math_functions() {
    std::cout << "\n=== Testing Math Functions ===\n";
    
    // Test Pow
    std::cout << "Pow(2.0, 3): " << std::fixed << std::setprecision(6) << Pow(2.0, 3) << std::endl;
    std::cout << "Pow(2.0, -2): " << Pow(2.0, -2) << std::endl;
    std::cout << "Pow(5.0, 0): " << Pow(5.0, 0) << std::endl;
    
    // Test Pow2RoundUp
    std::cout << "Pow2RoundUp(10): " << Pow2RoundUp(10) << std::endl;
    std::cout << "Pow2RoundUp(16): " << Pow2RoundUp(16) << std::endl;
    std::cout << "Pow2RoundUp(100): " << Pow2RoundUp(100) << std::endl;
    
    // Test Softmax
    std::vector<double> softmax_vec = {1.0, 2.0, 3.0};
    Softmax(&softmax_vec);
    std::cout << "Softmax([1.0, 2.0, 3.0]): ";
    for (double d : softmax_vec) {
        std::cout << std::setprecision(6) << d << " ";
    }
    std::cout << std::endl;
    
    // Test AvoidInf
    std::cout << "AvoidInf(1e301): " << AvoidInf(1e301) << std::endl;
    std::cout << "AvoidInf(-1e301): " << AvoidInf(-1e301) << std::endl;
    std::cout << "AvoidInf(NaN): " << AvoidInf(std::numeric_limits<double>::quiet_NaN()) << std::endl;
    
    // Test RoundInt
    std::cout << "RoundInt(3.7): " << RoundInt(3.7) << std::endl;
    std::cout << "RoundInt(3.2): " << RoundInt(3.2) << std::endl;
    std::cout << "RoundInt(-2.7): " << RoundInt(-2.7) << std::endl;
}

void test_bitset_functions() {
    std::cout << "\n=== Testing Bitset Functions ===\n";
    
    // Test EmptyBitset
    std::vector<uint32_t> bitset = EmptyBitset(100);
    std::cout << "EmptyBitset(100) size: " << bitset.size() << std::endl;
    
    // Test InsertBitset and FindInBitset
    InsertBitset(&bitset, 5);
    InsertBitset(&bitset, 15);
    InsertBitset(&bitset, 95);
    
    std::cout << "After inserting 5, 15, 95:" << std::endl;
    std::cout << "FindInBitset(bitset, 5): " << std::boolalpha << FindInBitset(bitset.data(), bitset.size(), 5) << std::endl;
    std::cout << "FindInBitset(bitset, 15): " << std::boolalpha << FindInBitset(bitset.data(), bitset.size(), 15) << std::endl;
    std::cout << "FindInBitset(bitset, 95): " << std::boolalpha << FindInBitset(bitset.data(), bitset.size(), 95) << std::endl;
    std::cout << "FindInBitset(bitset, 6): " << std::boolalpha << FindInBitset(bitset.data(), bitset.size(), 6) << std::endl;
    
    // Test ConstructBitset
    std::vector<int> vals = {1, 3, 7, 31, 63};
    std::vector<uint32_t> constructed_bitset = ConstructBitset(vals.data(), vals.size());
    std::cout << "ConstructBitset([1,3,7,31,63]) size: " << constructed_bitset.size() << std::endl;
    std::cout << "Contains 7: " << std::boolalpha << FindInBitset(constructed_bitset.data(), constructed_bitset.size(), 7) << std::endl;
    std::cout << "Contains 8: " << std::boolalpha << FindInBitset(constructed_bitset.data(), constructed_bitset.size(), 8) << std::endl;
}

void test_utility_functions() {
    std::cout << "\n=== Testing Utility Functions ===\n";
    
    // Test CheckDoubleEqualOrdered
    std::cout << "CheckDoubleEqualOrdered(1.0, 1.0): " << std::boolalpha << CheckDoubleEqualOrdered(1.0, 1.0) << std::endl;
    std::cout << "CheckDoubleEqualOrdered(1.0, 1.0000000001): " << std::boolalpha << CheckDoubleEqualOrdered(1.0, 1.0000000001) << std::endl;
    
    // Test GetDoubleUpperBound
    double upper = GetDoubleUpperBound(1.0);
    std::cout << "GetDoubleUpperBound(1.0): " << std::scientific << std::setprecision(17) << upper << std::endl;
    
    // Test GetLine
    std::cout << "GetLine('hello\\nworld'): " << GetLine("hello\nworld") << std::endl;
    std::cout << "GetLine('hello world'): " << GetLine("hello world") << std::endl;
    
    // Test CheckAllowedJSON
    std::cout << "CheckAllowedJSON('hello'): " << std::boolalpha << CheckAllowedJSON("hello") << std::endl;
    std::cout << "CheckAllowedJSON('hello{world}'): " << std::boolalpha << CheckAllowedJSON("hello{world}") << std::endl;
}

void test_statistics_functions() {
    std::cout << "\n=== Testing Statistics Functions ===\n";
    
    // Test ObtainMinMaxSum
    std::vector<double> data = {3.5, 1.2, 5.8, 2.1, 4.7};
    double min_val, max_val, sum_val;
    ObtainMinMaxSum(data.data(), data.size(), &min_val, &max_val, &sum_val);
    
    std::cout << "ObtainMinMaxSum([3.5, 1.2, 5.8, 2.1, 4.7]):" << std::endl;
    std::cout << "  Min: " << std::fixed << std::setprecision(2) << min_val << std::endl;
    std::cout << "  Max: " << max_val << std::endl;
    std::cout << "  Sum: " << sum_val << std::endl;
}

int main() {
    std::cout << "=== C++ LightGBM Common Functions Test ===\n";
    
    test_string_functions();
    test_numeric_functions();
    test_array_functions();
    test_math_functions();
    test_bitset_functions();
    test_utility_functions();
    test_statistics_functions();
    
    std::cout << "\n=== All tests completed ===\n";
    return 0;
}

File: test_common_equivalence.rs
use lightgbm_rust::core::utils::common::Common;

fn test_string_functions() {
    println!("\n=== Testing String Functions ===");
    
    // Test tolower
    println!("tolower('A'): {}", Common::tolower('A'));
    println!("tolower('Z'): {}", Common::tolower('Z'));
    println!("tolower('a'): {}", Common::tolower('a'));
    println!("tolower('1'): {}", Common::tolower('1'));
    
    // Test Trim
    println!("Trim('  hello  '): '{}'", Common::trim("  hello  ".to_string()));
    println!("Trim('\\t\\n hello \\r\\n'): '{}'", Common::trim("\t\n hello \r\n".to_string()));
    println!("Trim(''): '{}'", Common::trim("".to_string()));
    
    // Test RemoveQuotationSymbol
    println!("RemoveQuotationSymbol('\"hello\"'): '{}'", Common::remove_quotation_symbol("\"hello\"".to_string()));
    println!("RemoveQuotationSymbol(\"'world'\"): '{}'", Common::remove_quotation_symbol("'world'".to_string()));
    
    // Test StartsWith
    println!("StartsWith('hello world', 'hello'): {}", Common::starts_with("hello world", "hello"));
    println!("StartsWith('hello world', 'world'): {}", Common::starts_with("hello world", "world"));
    
    // Test Split
    let split_result = Common::split("a,b,c", ',');
    print!("Split('a,b,c', ','): ");
    for s in &split_result {
        print!("'{}' ", s);
    }
    println!();
    
    let split_result = Common::split("a,,b", ',');
    print!("Split('a,,b', ','): ");
    for s in &split_result {
        print!("'{}' ", s);
    }
    println!();
    
    // Test SplitLines
    let lines_result = Common::split_lines("line1\nline2\r\nline3");
    print!("SplitLines('line1\\nline2\\r\\nline3'): ");
    for s in &lines_result {
        print!("'{}' ", s);
    }
    println!();
}

fn test_numeric_functions() {
    println!("\n=== Testing Numeric Functions ===");
    
    // Test Atoi
    match Common::atoi::<i32>("123abc") {
        Ok((value, consumed)) => {
            let remaining = &"123abc"[consumed..];
            println!("Atoi('123abc'): value={}, remaining='{}'", value, remaining);
        }
        Err(_) => println!("Atoi('123abc'): failed to parse"),
    }
    
    match Common::atoi::<i32>("-456") {
        Ok((value, consumed)) => {
            let remaining = &"-456"[consumed..];
            println!("Atoi('-456'): value={}, remaining='{}'", value, remaining);
        }
        Err(_) => println!("Atoi('-456'): failed to parse"),
    }
    
    // Test Atof
    match Common::atof("123.45") {
        Ok((value, _)) => println!("Atof('123.45'): value={:.6}", value),
        Err(_) => println!("Atof('123.45'): failed to parse"),
    }
    
    match Common::atof("1.23e-4") {
        Ok((value, _)) => println!("Atof('1.23e-4'): value={:.6e}", value),
        Err(_) => println!("Atof('1.23e-4'): failed to parse"),
    }
    
    match Common::atof("nan") {
        Ok((value, _)) => println!("Atof('nan'): value={}", if value.is_nan() { "NaN".to_string() } else { value.to_string() }),
        Err(_) => println!("Atof('nan'): failed to parse"),
    }
    
    match Common::atof("inf") {
        Ok((value, _)) => println!("Atof('inf'): value={}", if value.is_infinite() { "Inf".to_string() } else { value.to_string() }),
        Err(_) => println!("Atof('inf'): failed to parse"),
    }
    
    // Test AtoiAndCheck
    println!("AtoiAndCheck('123'): {:?}", Common::atoi_and_check("123"));
    println!("AtoiAndCheck('123abc'): {:?}", Common::atoi_and_check("123abc"));
    
    // Test AtofAndCheck
    println!("AtofAndCheck('123.45'): {:?}", Common::atof_and_check("123.45"));
    println!("AtofAndCheck('123.45abc'): {:?}", Common::atof_and_check("123.45abc"));
}

fn test_array_functions() {
    println!("\n=== Testing Array Functions ===");
    
    // Test ArrayCast
    let int_vec = vec![1i32, 2, 3, 4, 5];
    let double_vec: Vec<f64> = Common::array_cast(&int_vec);
    print!("ArrayCast<int, double>([1,2,3,4,5]): ");
    for d in &double_vec {
        print!("{} ", d);
    }
    println!();
    
    // Test StringToArray
    let str_to_array: Vec<i32> = Common::string_to_array("1,2,3,4,5", ',');
    print!("StringToArray<int>('1,2,3,4,5', ','): ");
    for i in &str_to_array {
        print!("{} ", i);
    }
    println!();
    
    // Test Join
    let join_vec = vec![10, 20, 30, 40];
    let joined = Common::join(&join_vec, ", ");
    println!("Join([10,20,30,40], ', '): '{}'", joined);
    
    let joined_range = Common::join_range(&join_vec, 1, 3, " | ");
    println!("Join([10,20,30,40], 1, 3, ' | '): '{}'", joined_range);
}

fn test_math_functions() {
    println!("\n=== Testing Math Functions ===");
    
    // Test Pow
    println!("Pow(2.0, 3): {:.6}", Common::pow(2.0, 3));
    println!("Pow(2.0, -2): {:.6}", Common::pow(2.0, -2));
    println!("Pow(5.0, 0): {:.6}", Common::pow(5.0, 0));
    
    // Test Pow2RoundUp
    println!("Pow2RoundUp(10): {}", Common::pow2_round_up(10));
    println!("Pow2RoundUp(16): {}", Common::pow2_round_up(16));
    println!("Pow2RoundUp(100): {}", Common::pow2_round_up(100));
    
    // Test Softmax
    let mut softmax_vec = vec![1.0, 2.0, 3.0];
    Common::softmax(&mut softmax_vec);
    print!("Softmax([1.0, 2.0, 3.0]): ");
    for d in &softmax_vec {
        print!("{:.6} ", d);
    }
    println!();
    
    // Test AvoidInf
    println!("AvoidInf(1e301): {}", Common::avoid_inf_f64(1e301));
    println!("AvoidInf(-1e301): {}", Common::avoid_inf_f64(-1e301));
    println!("AvoidInf(NaN): {}", Common::avoid_inf_f64(f64::NAN));
    
    // Test RoundInt
    println!("RoundInt(3.7): {}", Common::round_int(3.7));
    println!("RoundInt(3.2): {}", Common::round_int(3.2));
    println!("RoundInt(-2.7): {}", Common::round_int(-2.7));
}

fn test_bitset_functions() {
    println!("\n=== Testing Bitset Functions ===");
    
    // Test EmptyBitset
    let mut bitset = Common::empty_bitset(100);
    println!("EmptyBitset(100) size: {}", bitset.len());
    
    // Test InsertBitset and FindInBitset
    Common::insert_bitset(&mut bitset, 5usize);
    Common::insert_bitset(&mut bitset, 15usize);
    Common::insert_bitset(&mut bitset, 95usize);
    
    println!("After inserting 5, 15, 95:");
    println!("FindInBitset(bitset, 5): {}", Common::find_in_bitset(&bitset, 5usize));
    println!("FindInBitset(bitset, 15): {}", Common::find_in_bitset(&bitset, 15usize));
    println!("FindInBitset(bitset, 95): {}", Common::find_in_bitset(&bitset, 95usize));
    println!("FindInBitset(bitset, 6): {}", Common::find_in_bitset(&bitset, 6usize));
    
    // Test ConstructBitset
    let vals = vec![1usize, 3, 7, 31, 63];
    let constructed_bitset = Common::construct_bitset(&vals);
    println!("ConstructBitset([1,3,7,31,63]) size: {}", constructed_bitset.len());
    println!("Contains 7: {}", Common::find_in_bitset(&constructed_bitset, 7usize));
    println!("Contains 8: {}", Common::find_in_bitset(&constructed_bitset, 8usize));
}

fn test_utility_functions() {
    println!("\n=== Testing Utility Functions ===");
    
    // Test CheckDoubleEqualOrdered
    println!("CheckDoubleEqualOrdered(1.0, 1.0): {}", Common::check_double_equal_ordered(1.0, 1.0));
    println!("CheckDoubleEqualOrdered(1.0, 1.0000000001): {}", Common::check_double_equal_ordered(1.0, 1.0000000001));
    
    // Test GetDoubleUpperBound
    let upper = Common::get_double_upper_bound(1.0);
    println!("GetDoubleUpperBound(1.0): {:.17e}", upper);
    
    // Test GetLine
    println!("GetLine('hello\\nworld'): {}", Common::get_line("hello\nworld"));
    println!("GetLine('hello world'): {}", Common::get_line("hello world"));
    
    // Test CheckAllowedJSON
    println!("CheckAllowedJSON('hello'): {}", Common::check_allowed_json("hello"));
    println!("CheckAllowedJSON('hello{{world}}'): {}", Common::check_allowed_json("hello{world}"));
}

fn test_statistics_functions() {
    println!("\n=== Testing Statistics Functions ===");
    
    // Test ObtainMinMaxSum
    let data = vec![3.5, 1.2, 5.8, 2.1, 4.7];
    let (min_val, max_val, sum_val): (f64, f64, f64) = Common::obtain_min_max_sum(&data);
    
    println!("ObtainMinMaxSum([3.5, 1.2, 5.8, 2.1, 4.7]):");
    println!("  Min: {:.2}", min_val);
    println!("  Max: {:.2}", max_val);
    println!("  Sum: {:.2}", sum_val);
}

fn main() {
    println!("=== Rust LightGBM Common Functions Test ===");
    
    test_string_functions();
    test_numeric_functions();
    test_array_functions();
    test_math_functions();
    test_bitset_functions();
    test_utility_functions();
    test_statistics_functions();
    
    println!("\n=== All tests completed ===");
}